pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,706 @@
+		ipFlow.SendBytes = entry.Forward.Bytes
+		ipFlow.RecvPkts = entry.Reverse.Packets
+		ipFlow.RecvBytes = entry.Reverse.Bytes
+	} else {
+		ipFlow.SendPkts = entry.Reverse.Packets
+		ipFlow.SendBytes = entry.Reverse.Bytes
+		ipFlow.RecvPkts = entry.Forward.Packets
+		ipFlow.RecvBytes = entry.Forward.Bytes
+	}
+	return ipFlow
+}
+
+func checkAppIPAddr(appAddrs []net.IP, entryIP net.IP) bool {
+	for _, appIP := range appAddrs {
+		if strings.Compare(appIP.String(), entryIP.String()) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// print the FlowTuple entries
+func (s *flowStats) String() string {
+	tstart := time.Unix(0, int64(s.TimeStart))
+	tout := int32(s.TimeOut)
+	return fmt.Sprintf("TS %v, TO %d(sec), proto %d src=%s dst=%s sport=%d dport=%d, snd=pkts/bytes %d/%d rcv=pkts/bytes %d/%d app-init %v, appNum %d, aclnum %d",
+		tstart, tout, s.Proto, s.SrcIP, s.DstIP, s.SrcPort, s.DstPort, s.SendPkts, s.SendBytes,
+		s.RecvPkts, s.RecvBytes, s.AppInitiate, s.appNum, s.aclNum)
+}
+
+func checkAppAndACL(ctx *zedrouterContext, instData networkAttrs) {
+	pub := ctx.pubAppNetworkStatus
+	items := pub.GetAll()
+	for _, st := range items {
+		status := cast.CastAppNetworkStatus(st)
+		for i, ulStatus := range status.UnderlayNetworkList {
+			log.Infof("===FlowStats: (index %d) AppNum %d, VifInfo %v, IP addr %v, Hostname %s\n",
+				i, status.AppNum, ulStatus.VifInfo, ulStatus.AllocatedIPAddr, ulStatus.HostName)
+
+			if ulStatus.AllocatedIPAddr != "" {
+				tmpIPs := net.ParseIP(ulStatus.AllocatedIPAddr)
+				instData.appIPAddrs[status.AppNum] = append(instData.appIPAddrs[status.AppNum], tmpIPs)
+			}
+
+			ulconfig := ulStatus.UnderlayNetworkConfig
+			intfAttr := aclAttr{
+				vif:     ulStatus.VifInfo.Vif,
+				bridge:  ulStatus.Bridge,
+				netUUID: ulconfig.Network,
+			}
+			instData.bnNet[ulStatus.Bridge] = intfAttr
+			if instData.appNet[status.AppNum] == nilUUID {
+				instData.appNet[status.AppNum] = status.UUIDandVersion.UUID
+				log.Infof("===FlowStats: appNet appNum %d, uuid %v\n", status.AppNum, instData.appNet[status.AppNum])
+			}
+
+			tmpMap := instData.ipaclattr[status.AppNum]
+			if tmpMap == nil {
+				tmpMap := make(map[int]aclAttr)
+				instData.ipaclattr[status.AppNum] = tmpMap
+			}
+			for _, rule := range ulStatus.ACLRules {
+				//log.Infof(" ==== FlowStats: ACL(%d), rule-ID %d, chain %s, aclname %s, prefix %v, rule %v, action %v\n",
+				//	j, rule.RuleID, rule.ActionChainName, rule.RuleName, rule.Prefix, rule.Rule, rule.Action)
+				var tempAttr aclAttr
+				tempAttr.aclNum = uint32(rule.RuleID)
+				tempAttr.chainName = rule.ActionChainName
+				tempAttr.aclName = rule.RuleName
+				tempAttr.tableName = rule.Table
+				tempAttr.vif = ulStatus.VifInfo.Vif
+				tempAttr.bridge = ulStatus.Bridge
+				if len(rule.Action) >= 2 { // '-j ACCEPT', '-j drop-all-bn1-nbu2x1'
+					tempAttr.action = rule.Action[1]
+				}
+				if _, ok := instData.ipaclattr[status.AppNum][int(rule.RuleID)]; !ok { // fake j as the aclNUM
```

Comment:
I guess you see these fakes because the user configured rule and it's corresponding marking rule will have the same rule id. That is how we maintain relationship between the two. I think you should only accept rules with "IsUserConfigured" variable of rule set to "true".

Reply:
ok, will add that